libbpf-rs/tests/bin/src/user_ringbuf.bpf.c

```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2024 Jose Fernandez
+
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+
+struct {
+	__uint(type, BPF_MAP_TYPE_USER_RINGBUF);
+	__uint(max_entries, 4096 /* one page */);
+} user_ringbuf SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, u32);
+	__type(value, u32);
+	__uint(max_entries, 100);
+} samples SEC(".maps");
+
+struct my_struct_t {
+	u32 key;
+	u32 value;
+};
+
+static long user_ringbuf_callback(struct bpf_dynptr *dynptr, void *context)
+{
+	const struct my_struct_t *data;
+
+	data = bpf_dynptr_data(dynptr, 0, sizeof(*data));
+	if (!data)
+		return 0;
+
+	bpf_map_update_elem(&samples, &data->key, &data->value, BPF_ANY);
+
+	return 0;
+}
+
+SEC("tp_btf/sched_switch")
```

Comment:
Let's use simple trigger on getpid syscall, as we do elsewhere, to not have everybody do their own thing for no reason?

```patch
diff --git tests/bin/src/user_ringbuf.bpf.c tests/bin/src/user_ringbuf.bpf.c
index debbf2..30d8fd 100644
--- tests/bin/src/user_ringbuf.bpf.c
+++ tests/bin/src/user_ringbuf.bpf.c
@@ -35,9 +35,8 @@ static long user_ringbuf_callback(struct bpf_dynptr *dynptr, void *context)
   return 0;
 }

-SEC("tp_btf/sched_switch")
-int BPF_PROG(handle__sched_switch, bool preempt, struct task_struct *prev,
-       struct task_struct *next)
+SEC("tp/syscalls/sys_enter_getpid")
+int handle__sys_enter_getpid(void *ctx)
 {
   bpf_user_ringbuf_drain(&user_ringbuf, user_ringbuf_callback, NULL, 0);

diff --git tests/bin/user_ringbuf.bpf.o tests/bin/user_ringbuf.bpf.o
index de04d5..40ac6a 100644
Binary files tests/bin/user_ringbuf.bpf.o and tests/bin/user_ringbuf.bpf.o differ
diff --git tests/test.rs tests/test.rs
index 50592a..7b487e 100644
--- tests/test.rs
+++ tests/test.rs
@@ -1205,7 +1205,7 @@ fn test_sudo_object_user_ringbuf() {

     let mut obj = get_test_object("user_ringbuf.bpf.o");
     let prog = obj
-        .prog_mut("handle__sched_switch")
+        .prog_mut("handle__sys_enter_getpid")
         .expect("failed to find program");
     let _link = prog.attach().expect("failed to attach prog");
     let urb_map = obj
@@ -1222,8 +1222,8 @@ fn test_sudo_object_user_ringbuf() {
         .submit(urb_sample)
         .expect("failed to submit sample");

-    // Sleep to ensure the sched_switch event is triggered
-    sleep(Duration::from_millis(10));
+    // Trigger BPF program.
+    let _pid = unsafe { libc::getpid() };

     // At this point, the BPF program should have run and consumed the sample in
     // the user ring buffer, and stored the key/value in the samples map.
```

or something like that?

Reply:
Changed hook to `sys_enter_getpid`.